### DISCUSSION TITLE
Accessing the `app` in `Depends`

### DISCUSSION THREAD

**Discussion by nicolaerosia** [fastapi/fastapi#6627] [Questions] [answered]
I would like to access the `FastAPI` app in a `Depends` function, is there a way to do that?
The rationale here is that I would like to avoid global variables and use the `state` in `app`.

Example usecase:
```
def get_session(app: fastapi.FastAPI) -> Session:
    [...]
    access the app.state here
    [...]

@router.post("/init")
def create(session: Session = Depends(get_session)):
[...]
use the SQLAlchemy session
[...]
```

**Comment by chbndrhnns** [comment]
`app` is available in the `request`. So would this work for you?

```
def get_session(request: Request) -> Session:
    assert request.app
```

**Comment by nicolaerosia** [comment]
@chbndrhnns I'm aware of that, but I would like to use Depends if it's possible.
For example, Depends makes it easy to use Authentication and in the Depends function I would like to use the state
```
oauth2_scheme = OAuth2PasswordBearer(tokenUrl="token")

def get_current_user(
    token: str = fastapi.Depends(oauth2_scheme),
) -> models.User:

**USE HERE SOMEHOW THE APP STATE**

[...]
get full details from SQL Database
validate
[...]
return models.User
```

**Comment by chbndrhnns** [comment, marked as answer]
This works for me:

```py
from fastapi import FastAPI, Depends
from starlette.requests import Request
from starlette.testclient import TestClient

app = FastAPI()


async def dep(request: Request):
    request.app.state.called = True


@app.get("/me", dependencies=[Depends(dep)])
async def me(request: Request):
    return {"called": request.app.state.called}


def test_dep():
    client = TestClient(app)
    res = client.get("/me")
    assert res.json()["called"]

```

**Comment by nicolaerosia** [comment]
Thank you very much, I couldn't find this in the docs, I tried it and it works!

**Comment by tiangolo** [comment]
Thanks for the help here @chbndrhnns ! 👏  🙇 

Thanks for reporting back and closing the issue @nicolaerosia  👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
